R/teal_data_utils.R

```diff
@@ -42,35 +42,37 @@ NULL
 .subset_teal_data <- function(data, datanames) {
   checkmate::assert_class(data, "teal_data")
   checkmate::assert_class(datanames, "character")
-  datanames_corrected <- intersect(datanames, ls(data@env))
-  dataname_corrected_with_raw <- intersect(c(datanames, sprintf("%s._raw_", datanames)), ls(data@env))
-
+  datanames_corrected <- intersect(datanames, .teal_data_ls(data))
+  datanames_corrected_with_raw <- c(datanames_corrected, ".raw_data")
```

Comment:
1. I'm thinking that this function should be part of `teal_data` class API.

2. I'm thinking about possible future extensions of this.
Can we make this `ls(envir = teal.code::get_env(data), pattern = "\\.", all.names = TRUE)` (plus change the var name).
Combining this with 1), `teal` (or anyone else) would be able to internal stuff with `.xyz` and `teal.data` would require any changes

Reply:
> I'm thinking that this function should be part of teal_data class API.

I have an idea that `@datanames` shouldn't exist and we should read the datanames (names) straight from the environment (except prefixed by `.`). It means that we can have `teal_data_ls` method registered in `teal.code`, with extension method `teal_data_ls.teal_data` which can handle topological sort (because join keys are teal_data property).

See the issue https://github.com/insightsengineering/teal.data/issues/333

> I'm thinking about possible future extensions of this.

I don't understand it yet.

Reply:
Please note also that dataname with `.` prefix is problematic. In `teal.slice` we have a `check_simple_name` which throws when dataname doesn't start with letter and doesn't contain only alphanumeric. [see here](https://github.com/insightsengineering/teal.slice/blob/aa05f7feb26bb32fbb5f069ec3e9e77d17450022/R/utils.R#L14)
This limitation is set because of modules (namespaces) are using datanames as an id. Dots will cause the problems with css selectors. Therefore I think dot-prefixed datanames should be avoided.

Reply:
> I have an idea that `@datanames` shouldn't exist

I'm all for it!

- no dataname slot - makes it simpler and reduce the clutter code made by end-user on the preprocessing part 
- topological sort functionality should live in the teal-data class

> It means that we can have teal_data_ls method registered in teal.code

One question tough. Can we avoid extending the base class in teal.code and apply extension only in teal.data? I would like to avoid edits in teal.code. It's good as is.

> > I'm thinking about possible future extensions of this.
> 
> I don't understand it yet.

I do imagine us having a need in the future to push yet another purely internal objects into the quenv. Right now we are excluding `.raw_data` value only. If we would exclude everything with dot upfront then this functionality would not require any edits. We would nicely follow open-close principle - open for extensions (i.e. "in order to create internal vars please do this with dot-name") but close for modifications (no need to further edits to support that).

> Please note also that dataname with . prefix is problematic.

Yes I am aware. It very connected to my old issue [here](https://github.com/insightsengineering/teal/issues/969). I guess we won't solve it here. But I think we should be good on this. I think that we are on the same page that `get_datanames()` functionality (however its implemented) should exclude datanames starting with dot. That check function you referenced should be called after that filter.

Reply:
responded in the issue https://github.com/insightsengineering/teal.data/issues/333#issuecomment-2376056555
I'm resolving this conversation here